python/dgl/merge.py

```diff
@@ -0,0 +1,103 @@
+"""Utilities for merging graphs."""
+
+import dgl
+from . import backend as F
+from .base import DGLError
+
+__all__ = ['merge']
+
+def merge(graphs):
+    r"""Merge a sequence of graphs together into a single graph.
+
+    Nodes and edges that exist in ``graphs[i+1]`` but not in ``dgl.merge(graphs[0:i+1])``
+    will be added to ``dgl.merge(graphs[0:i+1])`` along with their data.
+    Nodes that exist in both ``dgl.merge(graphs[0:i+1])`` and ``graphs[i+1]``
+    will be updated with ``graphs[i+1]``'s data if they do not match.
+
+    Parameters
+    ----------
+    graphs : list[DGLGraph]
+        Input graphs.
+
+    Notes
+    ----------
+    * Inplace updates are applied to a new, empty graph.
+    * Features that exist in ``dgl.graphs[i+1]`` will be created in
+      ``dgl.merge(dgl.graphs[i+1])`` if they do not already exist.
+
+    Examples
+    ----------
+
+    The following example uses PyTorch backend.
+
+    >>> import dgl
+    >>> import torch
+    >>> g = dgl.graph((torch.tensor([0,1]), torch.tensor([2,3])))
+    >>> g.ndata["x"] = torch.zeros(4)
+    >>> h = dgl.graph((torch.tensor([1,2]), torch.tensor([0,4])))
+    >>> h.ndata["x"] = torch.ones(5)
+    >>> m = dgl.merge([g,h])
+
+    ``m`` now contains edges and nodes from ``h`` and ``g``.
+    >>> m.edges()
+    (tensor([0, 1, 1, 2]), tensor([2, 3, 0, 4]))
+    >>> m.nodes()
+    tensor([0, 1, 2, 3, 4])
+
+    ``g``'s data has updated with ``h``'s in ``m``.
+    >>> m.ndata["x"]
+    tensor([1., 1., 1., 1., 1.])
+
+    See Also
+    ----------
+    add_nodes
+    add_edges
+    """
+
+    if len(graphs) == 0:
+        raise DGLError('The input list of graphs cannot be empty.')
+
+    ref = graphs[0]
+    ntypes = ref.ntypes
+    etypes = ref.canonical_etypes
+    data_dict = {etype: ([], []) for etype in etypes}
+    num_nodes_dict = {ntype: 0 for ntype in ntypes}
+    merged = dgl.heterograph(data_dict, num_nodes_dict, ref.idtype, ref.device)
+
+    # Merge edges and edge data.
+    for etype in etypes:
+        merged_us = F.zeros(0, ref.idtype, ref.device)
+        merged_vs = F.zeros(0, ref.idtype, ref.device)
+        edata_frames = []
+        for graph in graphs:
+            etype_id = graph.get_etype_id(etype)
+            us, vs = graph.edges(etype=etype)
+            merged_us = F.cat([merged_us, us], dim=0)
+            merged_vs = F.cat([merged_vs, vs], dim=0)
```

Comment:
It's best not to do recursive concatenation like:
```python
for x in xs:
    y = F.cat([y, x])
```
because concatenation is not an in-place operation and the above code will generate a lot of intermediate tensors and a long computation graph.

Instead, try gathering the small tensors in a list and call `F.cat` once.

Reply:
Thank you, good catch! I’ve implemented this change.